src/main/java/net/logstash/logback/composite/loggingevent/ThrowableSelectors.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2013-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.logstash.logback.composite.loggingevent;
+
+import ch.qos.logback.classic.spi.IThrowableProxy;
+
+public class ThrowableSelectors {
+
+    public static IThrowableProxy rootCause(IThrowableProxy throwable) {
+        // Keep a second pointer that slowly walks the causal chain.
+        // If the fast pointer ever catches the slower pointer, then there's a loop.
+        IThrowableProxy slowPointer = throwable;
+        boolean advanceSlowPointer = false;
+
+        IThrowableProxy cause;
+        while ((cause = throwable.getCause()) != null) {
+            throwable = cause;
+
+            if (throwable == slowPointer) {
+                // There's a cyclic reference, so no real root cause.
+                return null;
+            }
```

Comment:
Are cyclic references in cause chains really possible in real applications?  How can a cyclic reference be created (without using mocks)?  

For comparison, other frameworks (such as [spring](https://github.com/spring-projects/spring-framework/blob/7be68161297ec6f985750571d5289e90f8869c2a/spring-core/src/main/java/org/springframework/core/NestedExceptionUtils.java#L63-L74)) don't bother looking for cycles in cause chains.

Reply:
I hit that problem myself back in the day (although I forgot what caused it :see_no_evil:). One can create causal chains using `Throwable::initCause` quite easily.

As for references to other projects: Theres google/guava#2866 and [its implementation](https://github.com/google/guava/blob/v31.1/guava/src/com/google/common/base/Throwables.java#L255-L274) where I got the algorithm from. Logback also [added](https://github.com/qos-ch/logback/commit/6dbb8756186bc4a01ab7f22b8a9c0543311da00a) some cycle detection recently.

And, last but not least, Java itself does it, too:

```
jshell> var x = new Exception("x")
x ==> java.lang.Exception: x

jshell> var y = new Exception("y")
y ==> java.lang.Exception: y

jshell> x.initCause(y)
$3 ==> java.lang.Exception: x

jshell> y.initCause(x)
$4 ==> java.lang.Exception: y

jshell> x.printStackTrace()
java.lang.Exception: x
        at REPL.$JShell$11.do_it$Aux($JShell$11.java:6)
        at REPL.$JShell$11.do_it$($JShell$11.java:11)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at jdk.jshell/jdk.jshell.execution.DirectExecutionControl.invoke(DirectExecutionControl.java:209)
        at jdk.jshell/jdk.jshell.execution.RemoteExecutionControl.invoke(RemoteExecutionControl.java:116)
        at jdk.jshell/jdk.jshell.execution.DirectExecutionControl.invoke(DirectExecutionControl.java:119)
        at jdk.jshell/jdk.jshell.execution.ExecutionControlForwarder.processCommand(ExecutionControlForwarder.java:144)
        at jdk.jshell/jdk.jshell.execution.ExecutionControlForwarder.commandLoop(ExecutionControlForwarder.java:262)
        at jdk.jshell/jdk.jshell.execution.Util.forwardExecutionControl(Util.java:76)
        at jdk.jshell/jdk.jshell.execution.Util.forwardExecutionControlAndIO(Util.java:137)
        at jdk.jshell/jdk.jshell.execution.RemoteExecutionControl.main(RemoteExecutionControl.java:70)
Caused by: java.lang.Exception: y
        at REPL.$JShell$12.do_it$Aux($JShell$12.java:6)
        at REPL.$JShell$12.do_it$($JShell$12.java:11)
        ... 12 more
Caused by: [CIRCULAR REFERENCE: java.lang.Exception: x]
```

Reply:
Oh wow! Today I learned about the `initCause` method!   I've never seen that method used in 20+ years of programming Java.  I can't imagine why someone would want to use it, but I guess since it exists, we have to take it into consideration.

Thanks for enlightening me!